examples/benchmark.py

```diff
@@ -31,183 +124,469 @@ def downloadAmberSuite():
             cutoff = 2*unit.nanometers
-        if options.heavy:
-            dt = 0.005*unit.picoseconds
-            constraints = app.AllBonds
+        if bond_constraints == 'hbonds':
+            constraints = app.HBonds
+            hydrogenMass = 1.5*unit.amu
+        elif bond_constraints == 'allbonds':
+            constraints = app.AllBonds # JDC: Isn't constraining all bonds problematic?
             hydrogenMass = 4*unit.amu
-            if options.ensemble == 'NVE':
-                integ = mm.VerletIntegrator(dt)
-            else:
-                integ = mm.LangevinIntegrator(temperature, friction, dt)
         else:
+            raise ValueError(f"bond_constraints must be one of 'hbonds', 'allbonds': found {bond_constraints}")
+
+        positions = pdb.positions
+        system = ff.createSystem(pdb.topology, nonbondedMethod=method, nonbondedCutoff=cutoff, constraints=constraints, hydrogenMass=hydrogenMass)
+
+    return system, positions
+
+def serializeTest(directory=None, system=None, integrator=None, state=None, coredata=None, metadata=None):
+    """Serialize test system for benchmarking on Folding@home.
+
+    Parameters
+    ----------
+    directory : str
+        Directory to write serialized test XML files
+    system : openmm.System
+        The System to serialize
+    integrator : openmm.Integrator
+        The Integrator to serialize
+    state : openmm.State
+        The State to serialize
+    coredata : dict
+        Core parameters to serialize to XML
+    metadata : dict
+        Metadata to dump to YAML
+    """
+    import os
+    os.makedirs(directory, exist_ok=True)
+    import openmm
+    def serialize(obj, filename):
+        if filename.endswith('.bz2'):
+            import bz2
+            with bz2.open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+        elif filename.endswith('.gz'):
+            import gzip
+            with gzip.open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+        else:
+            with open(filename, 'wt') as outfile:
+                outfile.write(openmm.XmlSerializer.serialize(obj))
+
+    serialize(system, os.path.join(directory, 'system.xml.bz2'))
+    serialize(integrator, os.path.join(directory, 'integrator.xml.bz2'))
+    serialize(state, os.path.join(directory, 'state.xml.bz2'))
+
+    with open(os.path.join(directory, 'core.xml'), 'wt') as outfile:
+        outfile.write('<config>\n')
+        for key, value in coredata.items():
+            outfile.write(f'    <{key} v="{value}"/>\n')
+        outfile.write('</config>\n')
+
+    import yaml
+    with open(os.path.join(directory, 'metadata.yaml'), 'wt') as outfile:
+        outfile.write(yaml.dump(metadata))
+
+def runOneTest(testName, options):
+    """Perform a single benchmarking simulation."""
+
+    # Return if we do not support the requested precision
```

Comment:
I'd prefer to remove this check.  It makes it impossible to benchmark the Reference or CPU platform without manually specifying precision, which will be annoying in practice.  Let's just say that the precision option is ignored for these platforms.

Reply:
Removing this check would mean that benchmarks that run over multiple combinations of platform and precision will erroneously report that the Reference and CPU platforms were run in multiple precision modes (single, mixed, double) where the precision is actually ignored.

Reply:
You probably don't want to use a single run of the script for all options on all platforms.  Different platforms require different settings.  For example, the CPU platform generally should use a shorter PME cutoff than the GPU platforms.

And this check will be a serious annoyance for me.  I run this script constantly, often many times in a single day.  Having to add an unnecessary argument every time I do will be a pain.  It also will be very confusing to other users.  If you try to benchmark the Reference or CPU platform and don't explicitly specify a precision, it will simply exit without running any benchmarks and not explain why.

Reply:
> You probably don't want to use a single run of the script for all options on all platforms. Different platforms require different settings. For example, the CPU platform generally should use a shorter PME cutoff than the GPU platforms.

I'd love to move to a strategy where we autotune any platform-dependent optimizable parameters, but failing that, is there a way we could encode all these optimal combinations in some modular way?

Reply:
Autotuning is great as a future feature, but you have made a change that will make my life more difficult.  I probably run this script 10 times for every time anyone in the world other than me runs it.  The normal use case for this script is to run a single benchmark.  That will continue to be by far the most common way it gets run.  Please remove this check.

Reply:
I suggest we remove the ability to test multiple platforms on a single run of the script.  It's clear that creates more problems than it's worth.  There are too many differences between platforms to make that a sensible thing to do in most cases/